pkg/awsutils/imds.go

```diff
@@ -0,0 +1,216 @@
+package awsutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/pkg/errors"
+)
+
+// EC2MetadataIface is a subset of the EC2Metadata API.
+type EC2MetadataIface interface {
+	GetMetadataWithContext(ctx context.Context, p string) (string, error)
+}
+
+// typedIMDS is a typed wrapper around raw untyped IMDS SDK API.
+type TypedIMDS struct {
+	EC2MetadataIface
+}
+
+func (imds TypedIMDS) getList(ctx context.Context, key string) ([]string, error) {
+	data, err := imds.GetMetadataWithContext(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Fields(data), nil
+}
+
+// GetAZ returns the Availability Zone in which the instance launched.
+func (imds TypedIMDS) GetAZ(ctx context.Context) (string, error) {
+	return imds.GetMetadataWithContext(ctx, "placement/availability-zone")
+}
+
+// GetInstanceType returns the type of this instance.
+func (imds TypedIMDS) GetInstanceType(ctx context.Context) (string, error) {
+	return imds.GetMetadataWithContext(ctx, "instance-type")
+}
+
+// GetLocalIPv4 returns the private (primary) IPv4 address of the instance.
+func (imds TypedIMDS) GetLocalIPv4(ctx context.Context) (net.IP, error) {
+	return imds.getIP(ctx, "local-ipv4")
+}
+
+// GetInstanceID returns the ID of this instance.
+func (imds TypedIMDS) GetInstanceID(ctx context.Context) (string, error) {
+	return imds.GetMetadataWithContext(ctx, "instance-id")
+}
+
+// GetMac returns the first/primary network interface mac address.
+func (imds TypedIMDS) GetMac(ctx context.Context) (string, error) {
+	return imds.GetMetadataWithContext(ctx, "mac")
+}
+
+// GetMacs returns the interface addresses attached to the instance.
+func (imds TypedIMDS) GetMacs(ctx context.Context) ([]string, error) {
+	list, err := imds.getList(ctx, "network/interfaces/macs")
+	if err != nil {
+		return nil, err
+	}
+	// Remove trailing /
+	for i, item := range list {
+		list[i] = strings.TrimSuffix(item, "/")
+	}
+	return list, nil
+}
+
+// GetInterfaceID returns the ID of the network interface.
+func (imds TypedIMDS) GetInterfaceID(ctx context.Context, mac string) (string, error) {
+	key := fmt.Sprintf("network/interfaces/macs/%s/interface-id", mac)
+	return imds.GetMetadataWithContext(ctx, key)
+}
+
+func (imds TypedIMDS) getInt(ctx context.Context, key string) (int, error) {
+	data, err := imds.GetMetadataWithContext(ctx, key)
+	if err != nil {
+		return 0, err
+	}
+	return strconv.Atoi(data)
+}
+
+// GetDeviceNumber returns the unique device number associated with an interface.  The primary interface is 0.
```

Comment:
Is the primary interface always 0? I have kind of assumed it is, but is that documented somewhere?

Reply:
Not explicitly afaics :(  Primary ENI is assumed to be always "eth0" in all our scripts/docs/tutorials and the metadata docs for device-number say "eg eth3 is device-number 3" (*).  Primary ENI is also the first/only interface attached at boot, so assuming indices go up then the first is also always 0.

(*) .. which is obviously nonsense in the general case, because you can't know how the OS is going to do device enumeration - but I believe this is the fantasy that the original Xen API was based around.

I can remove this comment if you want.  The rest of our code already assumes 0==primary (eg [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/be9ca94f2b304b9359f4961d1e20e08d9aee5e94/pkg/awsutils/awsutils.go#L1063) and [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/be9ca94f2b304b9359f4961d1e20e08d9aee5e94/pkg/awsutils/awsutils.go#L562)) and this hasn't caused us any problems - yet.

Mostly I wanted to be more consistent across our codebase instead of using a mix of `.Primary` and `.DeviceIndex==0`, and here seemed a reasonable place to declare this assumption.  I can certainly soften it to say "vpc-cni assumes that primary interface is always 0"

Reply:
This is slightly different:
```
		// TODO: Verify that the first IP is always the primary
		primary := isFirst
```
This is checking with IP should not be used by pods. I've checked in the [aws/amazon-ec2-net-util](https://github.com/aws/amazon-ec2-net-utils/blob/f42f94acf498d21ef5e216fae56a5bd2601ea2a8/ec2net-functions#L97-L100) repo where they also use index 0 to be the "attchment IP". We should definitely see if we can get rid of this limitation though, and just reserve the "node IP".

The `aws.Int64Value(ec2res.Attachment.DeviceIndex) == 0` check to add that log line was to help customers who don't have `delete_on_termination` set for the primary ENI. #1023

Reply:
(Oh yes, you're right, one of my examples was primary *IP* not primary ENI)

So I'm happy to remove this, and instead make the code track primary ENI separately, and allow any DeviceNumber (including 0) for any ENI.

What I _don't_ want to do is have a function called `EC2InstanceMetadataCache.getENIDeviceNumber()` that does _not_ return the actual device number :P    Shall I put in a PR ahead of this one that tracks "primary-ness" explicitly?